Compression.md

```diff
@@ -0,0 +1,145 @@
+
+The compressor makes sure that we have all the data pertaining to a message accumulated. Subsequently, `zlib`'s deflater is invoked and deflated data is packed into a new WebSocketFrame, which is passed to the `WebSocketFrameEncoder`.
+
+#### 3.2 Decompressor implementation
+The decompressor is, functionally, a mirror image of the compressor. It is called `PermessageDeflateDecompressor` and is [ChannelInboundHandler](https://apple.github.io/swift-nio/docs/current/NIO/Protocols/ChannelInboundHandler.html).
+
+`ChannelInboundHandler`'s `channelRead(context:data)` method implemented here is invoked whenever a new `WebSocketFrame`s is produced by the `WebSocketDecoder`. Similar to the compressor, the decompressor processes data and continuation frames only. All the data pertaining to a message, possibly spread across continuation frames, is accumulated and `zlib`'s inflater is invoked. The inflated data is then packed into a new `WebSocketFrame` and moved into the next handler in the pipeline.
+
+#### 3.3 Configuring the Compressor and Decompressor
+[RFC7692](https://tools.ietf.org/html/rfc7692) defines four configuration options. They are actually two pairs of option, one for the client and server each:
+
+###### 3.3.1 `client_no_context_takeover` and `server_no_context_takeover` 
+These allow the client and server to use a new zlib inflater or deflater on every message. By default we reuse the inflater and deflater instances across messages. This means the inflater and deflater are initialized only once. The memory allocation/deallocation happen only once and the history of the deflate/inflate stream can be reused. [Here](https://tools.ietf.org/html/rfc7692#section-7.2.3.2) is an example.
+
+In the `Kitura-WebSocket-NIO` implementation:
+ - a client can inform the server that it isn't using context takeover by sending the `client_no_context_takeover` extension parameter. The server will respond with the same `client_no_context_takeover` parameter and configure its decompressor to not use context takeover.
+ - a client can request the server to not use context takeover by sending the `server_no_context_takeover` extension parameter. The server will regard this request and configure its compressor to not use context takeover. It will add the `server_no_context_takeover` parameter in the response.
+ - by default the server will configure its compressor and decompressor to use context takeover .i.e. to reuse compression context
+
+###### 3.3.2 `client_max_window_bits` and `server_max_window_bits` 
+These allow the client and server to share the LZ77 sliding window size. The default value is 15 bits which represents a window size of 32768 (2^15). It is necessary that the client's compressor and server's decompressor share the same LZ77 window size. So must it be between the server's compressor and client's decompressor.
```

Comment:
I might read better as "The client's compressor and the server's decompressor must have the same LZ77 window size. The same restriction applies to the server's compressor and the client's decompressor."

Reply:
I agree.